### DISCUSSION TITLE
Option to cache errors

### DISCUSSION THREAD

**Discussion by CreativeTechGuy** [TanStack/query#1772] [Ideas]
**Describe the bug**
Some errors from APIs are non-retryable and should be cached. When they are not, then the component will try to continuously re-make the call every time a component is mounted even though we know that it'll fail again. Depending on the structure of the app this could create infinite loops in the worst case or make tons of unnecessary API calls in the most likely case.

**To Reproduce**
1. Create a component that uses `useQuery` and returns an error.
2. Unmount that component and re-mount that component.
3. `useQuery` will retry and get the same error.

**Expected behavior**
There should be a query option to cache errors. Either a boolean to indicate that all errors should be cached using the normal rules, but also a function which is passed the error and can return true/false to indicate if that specific error should be cached. (So a user can differentiate between a non-retryable error and a retryable error.)

**Screenshots**
N/A

**Additional context**
N/A

**Comment by TkDodo** [comment]
You could catch your error in the queryFn and transform it to a resolved promise in case of an error you’d like to cache, and re-throw all other errors, but the error will then show up in data, so you’d have to model your structure accordingly.

you can also implement the retry function to prevent retries for certain errors.

converting this to a discussion because it’s a feature request, not a bug. We can definitely keep discussing there :)

**Reply by CreativeTechGuy** [reply]
Yeah that'd complicate the application layer since you'd need to check both `.data` and `.error` for the error. And `.isError` and similar properties would also not be accurate.

I understand there are workarounds given the current features, but this seems like it'd be better implemented in the library itself and is likely a far more common use-case than many of the features that this library already has.

**Reply by boschni** [reply]
What if you set retryOnMount to false?

**Reply by TkDodo** [reply]
hm, i haven't come across the scenario, so I'm having a hard time imagining when I would need that. The error is "cached" in the `error` property, until a refetch happens (the usual suspects: windowfocus, new mount, manual invalidate, ...). Until then, if your component re-renders, you will get the error from the "cache". if you want to turn of retries, there's an option for that, so I don't think we should mix these two things.

I'm not saying your issue is not common just because I haven't seen it, but there are more things to consider when it comes to how errors are handled. Consider the following scenario:
- fetch succeeds, you now have data and status: success
- re-fetch fails, you now are still in status: isSuccess, you have data _and_ error set

what should happen then if i have a "cached" error? It seems like you "just" want to disable the query altogether if you get a certain kind of error, and that's what the `enabled` option is for.

A light-weight wrapper over `useQuery` should do:

```
const [enabled, setEnabled] = useState(true)
return useQuery(key, fn, { enabled, onError: (error) => setEnabled(false)})
```

in this example, the query will never refetch again once you have an error - of course you can narrow it down to check for specific errors. But I don't think this is something the library can / should do out of the box.

**Reply by CreativeTechGuy** [reply]
Here's an example of one major issue that is caused by the constant refetching when the query throws an error.

https://codesandbox.io/s/angry-stallman-7h3l9?file=/src/App.js

Both a parent and child use the same query. These can be very far apart in the tree but in my example they are close for conciseness. The parent renders a loading status or the child depending on the state of the query. You could imagine that this logic would be used to determine which of many different things were rendered once loaded depending on the response. Then in this child it also needs that data so it uses that same hook. Well when the child mounts and the request had an error, it'll trigger a refetch (even when `refetchOnMount: false`). This refetch will re-render the parent, change the loading status back to true, thus unmounting the child and repeating the cycle.

This is obviously worst case scenario, but this problem would be solved if the same error that the parent received when calling the hook the first time would be cached for when the child (or any other component called that same hook).

**Reply by TkDodo** [reply]
Thanks for the example!

> Well when the child mounts and the request had an error, it'll trigger a refetch (even when refetchOnMount: false)

I think this is the _real_ issue - `refetchOnMount` should _probably_ be respected in that case. Disabling the query like I showed above should work though.

**Reply by boschni** [reply]
Would this be an option? https://codesandbox.io/s/gracious-christian-yqpo0

**Reply by CreativeTechGuy** [reply]
@TkDodo Your lightweight wrapper is similar to what I'm looking for except I still want to use the `useQuery` response even when disabled. I'd still need access to that error object and the status of that query. And furthermore, when another component across the app makes a change and invalidates the cache of this query, I'd like it to now retry as it's been cleared from the cache. In your example, I'd need a way to communicate to the `setEnabled` state that the cache has been cleared and re-enable the query.

Maybe the bug here is related to the `refetch*` config options? If I have `refetchOnMount: false` I'd expect that to not refetch at all on mount but it still does when there was an error.

----

To answer this specific question:
> what should happen then if i have a "cached" error?

I'd say it should behave the same as cached data. Any other component in the app will not refetch that same query and will instantly get the `.error` property and `.isError` set. Only when the cacheTime or staleTime is up, an option like `refetchOnWindowFocus` is set to true, or the cache is cleared for that key should it refetch. This is the exact behavior that `useQuery` has for `.data` today, I'd like the option to apply all of that behavior to `.error` conditionally too. :)

**Reply by TkDodo** [reply]
@boschni today I learned about `retryOnMount` 😅 👍

**Reply by CreativeTechGuy** [reply]
Can someone please document `retryOnMount`? I feel like the `QueryClient` configs are not well documented at all. That seems to do exactly what I need but that's totally news to me!

Can I make a request to document all of the QueryClient args here: https://react-query.tanstack.com/reference/QueryClient

**Reply by TkDodo** [reply]
Yes, please go ahead and PR it. It should go here: https://react-query.tanstack.com/reference/useQuery

Next to `refetchOnMount` or so. This is what the `types` have to say about it:

```
  /**
   * If set to `false`, the query will not be retried on mount if it contains an error.
   * Defaults to `true`.
   */
  retryOnMount?: boolean
```

**Reply by CreativeTechGuy** [reply]
I'll add that to my TODO list for tomorrow. It's the middle of the night for me.

Thanks for all of your help folks! Good to know that it already does what I'm looking for, just undocumented. 😂
